src/mixins/yield-erc/ERC20Vault.sol

```diff
@@ -0,0 +1,150 @@
+
+    /// @notice Emitted after a successful withdrawal.
+    /// @param owner The address that withdrew from the Vault.
+    /// @param to The destination for withdrawn tokens.
+    /// @param underlyingAmount The amount of underlying tokens that were withdrawn.
+    event Withdraw(address indexed from, address indexed to, uint256 underlyingAmount);
+
+    /// @notice Deposit a specific amount of underlying tokens.
+    /// @param to The address to receive shares corresponding to the deposit
+    /// @param underlyingAmount The amount of the underlying token to deposit.
+    function deposit(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+        // Determine the equivalent amount of shares and mint them.
+        _mint(to, shares);
+
+        emit Deposit(msg.sender, to, underlyingAmount);
+
+        // Transfer in underlying tokens from the user.
+        // This will revert if the user does not have the amount specified.
+        underlying.safeTransferFrom(msg.sender, address(this), underlyingAmount);
+
+        afterDeposit(underlyingAmount);
+    }
+
+    /// @notice Withdraw a specific amount of underlying tokens.
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param underlyingAmount The amount of underlying tokens to withdraw.
+    function withdraw(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+
+        // Determine the equivalent amount of shares and burn them.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shares);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    /// @notice Redeem a specific amount of shares for underlying tokens.
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param shareAmount The amount of shares to redeem for underlying tokens.
+    function redeem(address to, uint256 shareAmount) external virtual returns (uint256 underlyingAmount) {
+
+        // Determine the equivalent amount of underlying tokens.
+        underlyingAmount = shareAmount.fmul(exchangeRate(), baseUnit);
+
+        // Burn the provided amount of shares.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shareAmount);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    function beforeWithdraw(uint256 underlyingAmount) internal virtual {}
```

Comment:
the decision to have only `beforeWithdraw` and `afterDeposit` hooks seems arbitrary.
For example, if one wants to distribute some token reward based on user balance, then it should trigger it before deposit, and not after. (see, e.g., MakerDAO's `cropJoin`).

Reply:
I'm sure need not only before/after hooks but also granulate it for before/after burn/mint and before/after token transfer.
It can be used in fees logic and determination correct vault PPFS in strategies.